cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,688 @@
+                        chunk_count++;
+                      });
+                  });
+                });
+
+  // Do we have any bloom filters
+  if (have_bloom_filters) {
+    // Vector to hold bloom filter data
+    std::vector<rmm::device_buffer> bloom_filter_data(num_chunks);
+
+    // Read bloom filter data
+    read_bloom_filter_data(sources,
+                           num_chunks,
+                           bloom_filter_data,
+                           bloom_filter_offsets,
+                           bloom_filter_sizes,
+                           chunk_source_map,
+                           stream);
+
+    // Return bloom filter data
+    return bloom_filter_data;
+  }
+
+  // Return empty vector
+  return {};
+}
+
+std::vector<Type> aggregate_reader_metadata::get_parquet_types(
+  host_span<std::vector<size_type> const> row_group_indices,
+  host_span<int const> column_schemas) const
+{
+  std::vector<Type> parquet_types(column_schemas.size());
+  // Find a source with at least one row group
+  auto const src_iter = std::find_if(row_group_indices.begin(),
+                                     row_group_indices.end(),
+                                     [](auto const& rg) { return rg.size() > 0; });
+  CUDF_EXPECTS(src_iter != row_group_indices.end(), "");
+
+  // Source index
+  auto const src_index = std::distance(row_group_indices.begin(), src_iter);
+  std::transform(column_schemas.begin(),
+                 column_schemas.end(),
+                 parquet_types.begin(),
+                 [&](auto const schema_idx) {
+                   // Use the first row group in this source
+                   auto constexpr row_group_index = 0;
+                   return get_column_metadata(row_group_index, src_index, schema_idx).type;
+                 });
+
+  return parquet_types;
+}
+
+std::optional<std::vector<std::vector<size_type>>> aggregate_reader_metadata::apply_bloom_filters(
+  host_span<std::unique_ptr<datasource> const> sources,
+  host_span<std::vector<size_type> const> input_row_group_indices,
+  host_span<data_type const> output_dtypes,
+  host_span<int const> output_column_schemas,
+  std::reference_wrapper<ast::expression const> filter,
+  rmm::cuda_stream_view stream) const
+{
+  // Number of input table columns
+  auto const num_input_columns = static_cast<cudf::size_type>(output_dtypes.size());
+
+  // Total number of row groups after StatsAST filtration
+  auto const total_row_groups = std::accumulate(
+    input_row_group_indices.begin(),
+    input_row_group_indices.end(),
+    size_t{0},
+    [](size_t sum, auto const& per_file_row_groups) { return sum + per_file_row_groups.size(); });
+
+  // Check if we have less than 2B total row groups.
+  CUDF_EXPECTS(total_row_groups <= std::numeric_limits<cudf::size_type>::max(),
+               "Total number of row groups exceed the size_type's limit");
+
+  auto mr = cudf::get_current_device_resource_ref();
```

Comment:
Are we using memory resources properly?

1. Let's define `mr` closer to where it is used
2. This returns data to the caller that is allocated with a memory resource that wasn't passed in. Do we need to accept an `mr` parameter in this function and use that for the data returned to the caller? I can't recall how this rule applies to detail functions if the memory is not returned to the _user_ but only an _internal caller_. https://github.com/rapidsai/cudf/blob/branch-25.02/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#output-memory_

Reply:
After submitting this review, I scrolled a little further in the link I posted. It does clarify that detail APIs should accept an `mr` parameter and use that for data returned to the caller.

> This rule automatically applies to all detail APIs that allocate memory. Any detail API may be called by any public API, and therefore could be allocating memory that is returned to the user. To support such uses cases, all detail APIs allocating memory resources should accept an `mr` parameter. Callers are responsible for either passing through a provided `mr` or `cudf::get_current_device_resource_ref()` as needed.

Reply:
I think we are okay here as we aren't returning any data allocated with `mr` in `apply_bloom_filter()` and sub-functions, and only using the `cudf::get_current_device_resource_ref();` to allocate temporary buffers which should be automatically when we return.

Reply:
I have removed the `mr` input parameter from all functions in this file and replaced with `cudf::get_current_device_resource_ref()` to allocate temp memory where needed